lightning/src/onion_message/messenger.rs

```diff
@@ -357,6 +371,103 @@ where
 		}
 	}
 
+	/// Decode one layer of an incoming onion message
+	/// Returns either a Forward (another onion message), or Receive (decrypted content)
+	pub fn peel_onion<T: CustomOnionMessageContents>(
+		node_signer: &NS,
+		secp_ctx: &Secp256k1<secp256k1::All>,
+		logger: &L, 
+		custom_handler: &CMH,
+		msg: &msgs::OnionMessage,
+	) -> Result<PeeledOnion<CMH>, ReceiveError> {
+		let control_tlvs_ss = match node_signer.ecdh(Recipient::Node, &msg.blinding_point, None) {
+			Ok(ss) => ss,
+			Err(e) =>  {
+				log_error!(logger, "Failed to retrieve node secret: {:?}", e);
+				return Err(ReceiveError {});
+			}
+		};
+		let onion_decode_ss = {
+			let blinding_factor = {
+				let mut hmac = HmacEngine::<Sha256>::new(b"blinded_node_id");
+				hmac.input(control_tlvs_ss.as_ref());
+				Hmac::from_engine(hmac).into_inner()
+			};
+			match node_signer.ecdh(Recipient::Node, &msg.onion_routing_packet.public_key,
+				Some(&Scalar::from_be_bytes(blinding_factor).unwrap()))
+			{
+				Ok(ss) => ss.secret_bytes(),
+				Err(()) => {
+					log_trace!(logger, "Failed to compute onion packet shared secret");
+					return Err(ReceiveError {});
+				}
+			}
+		};
+		match onion_utils::decode_next_untagged_hop(
+			onion_decode_ss, &msg.onion_routing_packet.hop_data[..], msg.onion_routing_packet.hmac,
+			(control_tlvs_ss, custom_handler.deref(), logger.deref())
+		) {
+			Ok((Payload::Receive::<<<CMH as Deref>::Target as CustomOnionMessageHandler>::CustomMessage> {
+				message, control_tlvs: ReceiveControlTlvs::Unblinded(ReceiveTlvs { path_id }), reply_path,
+			}, None)) => {
+				log_trace!(logger,
+					"Received an onion message with path_id {:02x?} and {} reply_path",
+						path_id, if reply_path.is_some() { "a" } else { "no" });
```

Comment:
I think this log should move to the receive case in `handle_onion_message` since LDK users calling this method will see the result anyway

Reply:
good call